src/Runner/TestSuiteSorter.php

```diff
@@ -62,11 +106,29 @@ private function sort(TestSuite $suite, int $order, bool $resolveDependencies):
             $suite->setTests($this->randomize($suite->tests()));
         }
 
+        if ($orderDefects === self::ORDER_DEFECTS_FIRST && $this->cache !== null) {
+            $suite->setTests($this->sortDefectsFirst($suite->tests()));
+        }
+
         if ($resolveDependencies && $suite->tests()[0] instanceof TestCase) {
             $suite->setTests($this->resolveDependencies($suite->tests()));
         }
     }
 
+    private function addSuiteToDefectSortOrder(TestSuite $suite): void
```

Comment:
**MAJOR**
This method can be called even when `$this->cache = null`, and then it gonna crash hardly.
It means whenever one want to call `addSuiteToDefectSortOrder`, he need to check `$this->cache !== null` manually, and he need to be aware about this! This proves that architecture of this class is defective.

Also, having a need to do that check everywhere isn't optimal - I would suggest to remove possibility to have that field nullable - eg, by introducing `TestResultCacheInterface` and `NullTestResultCache`

Reply:
Yup, I overlooked the check for the cache in `addSuiteToDefectSortOrder` since it would never happen internally. However it is good to check for it inside the method. At least for now I have added that.

The interface suggestion sounds like an interesting approach, I will give it a spin

Reply:
Nevermind the previous comment, I have given the interface a try. Is this the kind of pattern you had in mind?

Reply:
indeed